src/EventListener/VersionListener.php

```diff
@@ -49,32 +46,22 @@ public function postPersist(Version $version, LifecycleEventArgs $event): void
         $this->getEM()->getRepository(AuditRecord::class)->insert($record);
     }
 
-    public function preUpdate(Version $version, PreUpdateEventArgs $event): void
+    public function onVersionReferenceChanged(VersionReferenceChangedEvent $event): void
     {
-        if (($event->hasChangedField('source') || $event->hasChangedField('dist')) && !$version->isDevelopment()) {
-            $oldDistRef = $event->hasChangedField('dist') ? ($event->getOldValue('dist')['reference'] ?? null) : $version->getDist()['reference'] ?? null;
-            $oldSourceRef = $event->hasChangedField('source') ? ($event->getOldValue('source')['reference'] ?? null) : $version->getSource()['reference'] ?? null;
-            $newDistRef = $event->hasChangedField('dist') ? ($event->getNewValue('dist')['reference'] ?? null) : $version->getDist()['reference'] ?? null;
-            $newSourceRef = $event->hasChangedField('source') ? ($event->getNewValue('source')['reference'] ?? null) : $version->getSource()['reference'] ?? null;
-            if ($oldDistRef !== $newDistRef || $oldSourceRef !== $newSourceRef) {
-                // buffering things to be inserted in postUpdate once we can confirm it is done
-                $this->buffered[] = AuditRecord::versionReferenceChange($version, $oldSourceRef, $oldDistRef);
-            }
-        }
-    }
+        $version = $event->getVersion();
 
-    /**
-     * @param LifecycleEventArgs<EntityManager> $event
-     */
-    public function postUpdate(Version $version, LifecycleEventArgs $event): void
-    {
-        if ($this->buffered) {
-            $repo = $this->getEM()->getRepository(AuditRecord::class);
-            foreach ($this->buffered as $record) {
-                $repo->insert($record);
-            }
-            $this->buffered = [];
+        if ($version->isDevelopment() && !$event->hasMetadataChanged()) {
+            return;
         }
+
+        $record = AuditRecord::versionReferenceChange(
+            $version,
+            $event->getOldSourceRef(),
+            $event->getOldDistRef(),
+            $event->getNewMetadata()
+        );
+
+        $this->getEM()->getRepository(AuditRecord::class)->insert($record);
```

Comment:
Can we just persist these new records and let them be flushed to DB together with the new version entities? To ensure data integrity.

Reply:
Alright, makes sense too. Changed it!